File: 3152-maximum-value-of-an-ordered-triplet-ii/3152-maximum-value-of-an-ordered-triplet-ii.py
class Solution:
    def maximumTripletValue(self, nums: List[int]) -> int:
        n = len(nums)
        ans = dmax = imax = 0
        for i in range(n):
            ans = max(ans, dmax * nums[i])
            dmax = max(dmax, imax - nums[i])
            imax = max(imax, nums[i])
        
        return ans

File: 3152-maximum-value-of-an-ordered-triplet-ii/3152-maximum-value-of-an-ordered-triplet-ii.js
 /* @param {number[]} nums
 * @return {number}
 */
var maximumTripletValue = function (nums) {
    let n = nums.length;
    var ans = 0, dmax = 0, imax = 0;
    for (let num of nums) {
        ans = Math.max(ans, dmax * num);
        dmax = Math.max(dmax, imax - num);
        imax = Math.max(imax, num);
    }
    return ans;
};